spec/rubocop/cop/rspec/implicit_block_expectation_spec.rb

```diff
@@ -0,0 +1,127 @@
+  subject(:cop) { described_class.new }
+
+  it 'flags lambda in subject' do
+    expect_offense(<<-RUBY)
+      describe do
+        subject { -> { boom } }
+        it { is_expected.to change { something }.to(new_value) }
+             ^^^^^^^^^^^ Avoid implicit block expectations.
+      end
+    RUBY
+  end
+
+  it 'ignores non-lambda subject' do
+    expect_no_offenses(<<-RUBY)
+      describe do
+        subject { 'normal' }
+        it { is_expected.to eq(something) }
+      end
+    RUBY
+  end
+
+  it 'flags lambda in subject!' do
+    expect_offense(<<-RUBY)
+      describe do
+        subject! { -> { boom } }
+        it { is_expected.to change { something }.to(new_value) }
+             ^^^^^^^^^^^ Avoid implicit block expectations.
+      end
+    RUBY
+  end
+
+  it 'flags literal lambda' do
+    expect_offense(<<-RUBY)
+      describe do
+        subject! { lambda { boom } }
+        it { is_expected.to change { something }.to(new_value) }
+             ^^^^^^^^^^^ Avoid implicit block expectations.
+      end
+    RUBY
+  end
+
+  it 'flags proc' do
+    expect_offense(<<-RUBY)
+      describe do
+        subject! { proc { boom } }
+        it { is_expected.to change { something }.to(new_value) }
+             ^^^^^^^^^^^ Avoid implicit block expectations.
+      end
+    RUBY
+  end
+
+  it 'flags Proc.new' do
+    expect_offense(<<-RUBY)
+      describe do
+        subject! { Proc.new { boom } }
+        it { is_expected.to change { something }.to(new_value) }
+             ^^^^^^^^^^^ Avoid implicit block expectations.
+      end
+    RUBY
+  end
+
+  it 'flags named subject' do
+    expect_offense(<<-RUBY)
+      describe do
+        subject(:name) { -> { boom } }
+        it { is_expected.to change { something }.to(new_value) }
+             ^^^^^^^^^^^ Avoid implicit block expectations.
+      end
+    RUBY
+  end
+
+  it 'flags when subject is defined in the outer example group' do
+    expect_offense(<<-RUBY)
+      describe do
+        subject { -> { boom } }
+        context do
+          it { is_expected.to change { something }.to(new_value) }
+               ^^^^^^^^^^^ Avoid implicit block expectations.
+        end
+      end
+    RUBY
+  end
+
+  it 'ignores normal local subject' do
+    expect_no_offenses(<<-RUBY)
+      describe do
+        subject { -> { boom } }
+        context do
+          subject { 'normal' }
+          it { is_expected.to eq(something) }
+        end
+      end
+    RUBY
+  end
+
+  it 'ignores named subject with deeply nested lambda' do
+    expect_no_offenses(<<-RUBY)
+      describe do
+        subject { {hash: -> { boom }} }
+        it { is_expected.to be(something) }
+      end
+    RUBY
+  end
+
+  it 'flags with `should` as implicit subject' do
+    expect_offense(<<-RUBY)
+      describe do
+        subject { -> { boom } }
+        it { should change { something }.to(new_value) }
+             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Avoid implicit block expectations.
```

Comment:
The range is not ideal, but I decided to avoid complicating the cop for the sake of highlighting `should` itself.
The difference is that `is_expected` is a self-contained `send`, while for `should` it's the whole `should(change { something }.to(new_value))`.